python/orca/tutorial/tensorflow/ncf_orca.py

```diff
@@ -0,0 +1,155 @@
+                  metrics=['accuracy'])
+    return model
+
+cluster_mode = str(sys.argv[1])
+# cluster_mode = "k8s"
+if cluster_mode == "local":
+    sc = init_orca_context(memory=str(sys.argv[2]))
+elif cluster_mode == "standalone":
+    sc = init_orca_context("standalone", cores=int(sys.argv[3]), num_nodes=int(sys.argv[4]), memory=str(sys.argv[2]))
+elif cluster_mode == "yarn":
+    sc = init_orca_context("yarn", cores=int(sys.argv[3]), num_nodes=int(sys.argv[4]), memory=str(sys.argv[2]))
+elif cluster_mode == "k8s":
+    sc = init_orca_context(cluster_mode="k8s", cores=8, num_nodes=4, memory="20g",
+                           master="k8s://https://172.16.0.200:6443",
+                           container_image="10.239.45.10/arda/intelanalytics/bigdl-k8s-spark-3.1.2:0.14.0-SNAPSHOT",
+                           conf={"spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName": "nfsvolumeclaim",
+                                 "spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path": "/bigdl2.0/data",
+                                 "spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName": "nfsvolumeclaim",
+                                 "spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path": "/bigdl2.0/data"})
+elif cluster_mode == "spark-submit":  # To test k8s using spark-submit
+    sc = init_orca_context(cluster_mode="spark-submit")
+
+
+data_path = "./bigdl2.0/data"
+data_type = "ml-1m"
+# Need spark3 to support delimiter with more than one character.
+spark= OrcaContext.get_spark_session()
+from pyspark.sql.types import StructField, StructType, IntegerType
+
+schema = StructType(
+    [
+        StructField('user', IntegerType(), True),
+        StructField('item', IntegerType(), True)
+    ]
+)
+df = spark.read.csv("{}/{}/ratings.csv".format(data_path, data_type), sep="::",schema=schema,
+                     header=False)
+min_user_id = df.agg({"user": "min"}).collect()[0]["min(user)"]
+max_user_id = df.agg({"user": "max"}).collect()[0]["max(user)"]
+min_item_id = df.agg({"item": "min"}).collect()[0]["min(item)"]
+max_item_id = df.agg({"item": "max"}).collect()[0]["max(item)"]
+print(min_user_id, max_user_id, min_item_id, max_item_id)
+from pyspark.sql import functions
+df = df.withColumn('label', functions.lit(1))
+
+from pyspark.sql import functions as F
+from pyspark.sql import types as T
+import random
+
+neg_scale = 4
+def neg_sample(x):
+    item_count = len(x) * neg_scale
+    max_count = max_item_id - len(set(x))
+    neg_item = random.sample(set(range(min_item_id, max_item_id+1)) - set(x), min(item_count,max_count))
+    return neg_item
+
+neg_sample_udf = F.udf(neg_sample, T.ArrayType(IntegerType(), False))
+
+df_neg= df.groupBy('user').agg(neg_sample_udf(F.collect_list('item')).alias('item_list'))
+from pyspark.sql.functions import *
+df_neg = df_neg.select(df_neg.user, explode(df_neg.item_list))
```

Comment:
will the column have name item?

Reply:
> 

it won't have name, but its name is not important, because it will be unionAll after df